ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/CommandDeduplication.scala

```diff
@@ -3,90 +3,141 @@
             )
+        val dedupKey = commandDedupKey(transactionEntry.submitterInfo)
+        val dedupEntry = commitContext.get(dedupKey)
+        val maybeDedupValue = dedupEntry
+          .filter(_.hasCommandDedup)
+          .map(_.getCommandDedup)
+        val recordTimeOrMinimumRecordTime = commitContext.recordTime match {
+          case Some(recordTime) =>
+            // During the normal execution, in the deduplication state value we stored the record time
+            // This allows us to compare the record times directly
+            recordTime.toInstant
+          case None =>
+            // We select minimum record time for pre-execution
+            // During pre-execution in the deduplication state value we stored the maximum record time
+            // To guarantee the deduplication duration, we basically compare the maximum record time of the previous transaction
+            // with the minimum record time of the current transaction. This gives us the smallest possible interval between two transactions.
+            commitContext.minimumRecordTime
+              .getOrElse(
+                throw Err.InternalError("Minimum record time is not set for pre-execution")
+              )
+              .toInstant
+        }
+        val isDuplicate = maybeDedupValue
+          .flatMap(commandDeduplication =>
+            commandDeduplication.getTimeCase match {
+              // Backward-compatibility, will not  be set for new entries
+              case TimeCase.DEDUPLICATED_UNTIL =>
+                Some(parseTimestamp(commandDeduplication.getDeduplicatedUntil))
+              // Set during normal execution, no time skews are added
+              case TimeCase.RECORD_TIME =>
+                val storedDuplicateRecordTime =
+                  parseTimestamp(commandDeduplication.getRecordTime)
+                Some(
+                  storedDuplicateRecordTime
+                    .add(commandDeduplicationDuration)
+                )
+              // Set during pre-execution, time skews are already accounted for
+              case TimeCase.MAX_RECORD_TIME =>
+                val maxRecordTime =
+                  parseTimestamp(commandDeduplication.getMaxRecordTime)
+                Some(
+                  maxRecordTime
+                    .add(commandDeduplicationDuration)
+                )
+              case TimeCase.TIME_NOT_SET =>
+                None
+            }
+          )
+          .forall(deduplicatedUntil =>
+            recordTimeOrMinimumRecordTime.isAfter(deduplicatedUntil.toInstant)
+          )
+        if (isDuplicate) {
+          StepContinue(transactionEntry)
+        } else {
+          if (commitContext.preExecute) {
+            // The out of time bounds entry is required in the committer, so we set it to the default value as we stop the steps here with the duplicate rejection
+            commitContext.outOfTimeBoundsLogEntry = Some(DamlLogEntry.getDefaultInstance)
```

Comment:
This is not going to get consumed on the read path because we won't be setting record time bounds in this case, right?

Reply:
That is right, as we write the successful write set in this case, therefore this is just for the purpose of passing committer validation.

Reply:
Thanks for the confirmation. It may be worth adding a note that we assume the read path is going to handle an empty `DamlLogEntry`.